File: backend/app/services/kyc_provider.py
from abc import ABC, abstractmethod
from typing import Dict, Any

class KycProvider(ABC):
    @abstractmethod
    def verify_pan(self, pan_number: str) -> Dict[str, Any]:
        pass

    @abstractmethod
    def verify_aadhaar(self, aadhaar_number: str, otp: str) -> Dict[str, Any]:
        pass

    @abstractmethod
    def link_bank(self, account_id: str) -> Dict[str, Any]:
        pass

    @abstractmethod
    def fetch_cibil(self, user_id: str) -> Dict[str, Any]:
        pass

class MockKycProvider(KycProvider):
    def verify_pan(self, pan_number: str) -> Dict[str, Any]:
        clean_pan = pan_number.upper() if pan_number else "ABCPS1234F"
        return {
            "pan_number": clean_pan,
            "full_name": "RAMESH KUMAR SHARMA",
            "date_of_birth": "15/08/1988",
            "status": "VALID_ACTIVE",
            "is_demo_data": True
        }

    def verify_aadhaar(self, aadhaar_number: str, otp: str) -> Dict[str, Any]:
        last4 = aadhaar_number[-4:] if len(aadhaar_number) >= 4 else "4892"
        return {
            "masked_aadhaar": f"XXXX-XXXX-{last4}",
            "name": "Ramesh Kumar Sharma",
            "address": "Village Shivpur, District Varanasi, Uttar Pradesh - 221003",
            "is_verified": len(otp) == 6,
            "is_demo_data": True
        }

    def link_bank(self, account_id: str) -> Dict[str, Any]:
        return {
            "account_id": account_id or "ACC-9048102",
            "bank_name": "Bank of Baroda (Rural Shivpur Branch)",
            "account_number_masked": "XXXX-XXXX-4819",
            "ifsc_code": "BARB0VASHIV",
            "is_aa_linked": True,
            "avg_monthly_balance": 14850.0,
            "is_demo_data": True
        }

    def fetch_cibil(self, user_id: str) -> Dict[str, Any]:
        return {
            "score": 742,
            "total_accounts": 3,
            "active_loans": 1,
            "on_time_repayment_percent": 96.5,
            "risk_category": "Low Risk (Good Standing)",
            "is_demo_data": True
        }

File: app/src/main/java/com/example/data/service/KycProvider.kt
package com.example.data.service

data class PanDetails(
    val panNumber: String,
    val fullName: String,
    val dateOfBirth: String,
    val status: String,
    val isDemoData: Boolean = true
)

data class AadhaarDetails(
    val maskedAadhaar: String,
    val name: String,
    val address: String,
    val isVerified: Boolean,
    val isDemoData: Boolean = true
)

data class BankAccount(
    val accountId: String,
    val bankName: String,
    val accountNumberMasked: String,
    val ifscCode: String,
    val isAaLinked: Boolean,
    val avgMonthlyBalance: Double,
    val isDemoData: Boolean = true
)

data class CibilReport(
    val score: Int,
    val totalAccounts: Int,
    val activeLoans: Int,
    val onTimeRepaymentPercent: Double,
    val riskCategory: String,
    val isDemoData: Boolean = true
)

interface KycProvider {
    suspend fun verifyPan(panNumber: String): PanDetails
    suspend fun verifyAadhaar(aadhaarNumber: String, otp: String): AadhaarDetails
    suspend fun linkBank(accountId: String): BankAccount
    suspend fun fetchCibil(userId: String): CibilReport
}

class MockKycProvider : KycProvider {
    override suspend fun verifyPan(panNumber: String): PanDetails {
        val cleanPan = if (panNumber.isNotBlank()) panNumber.uppercase() else "ABCPS1234F"
        return PanDetails(
            panNumber = cleanPan,
            fullName = "RAMESH KUMAR SHARMA",
            dateOfBirth = "15/08/1988",
            status = "VALID_ACTIVE",
            isDemoData = true
        )
    }

    override suspend fun verifyAadhaar(aadhaarNumber: String, otp: String): AadhaarDetails {
        val last4 = if (aadhaarNumber.length >= 4) aadhaarNumber.takeLast(4) else "4892"
        return AadhaarDetails(
            maskedAadhaar = "XXXX-XXXX-$last4",
            name = "Ramesh Kumar Sharma",
            address = "Village Shivpur, District Varanasi, Uttar Pradesh - 221003",
            isVerified = otp.isNotBlank() && otp.length == 6,
            isDemoData = true
        )
    }

    override suspend fun linkBank(accountId: String): BankAccount {
        return BankAccount(
            accountId = accountId.ifBlank { "ACC-9048102" },
            bankName = "Bank of Baroda (Rural Shivpur Branch)",
            accountNumberMasked = "XXXX-XXXX-4819",
            ifscCode = "BARB0VASHIV",
            isAaLinked = true,
            avgMonthlyBalance = 14850.0,
            isDemoData = true
        )
    }

    override suspend fun fetchCibil(userId: String): CibilReport {
        return CibilReport(
            score = 742,
            totalAccounts = 3,
            activeLoans = 1, // e.g. PM SVANidhi or KCC loan
            onTimeRepaymentPercent = 96.5,
            riskCategory = "Low Risk (Good Standing)",
            isDemoData = true
        )
    }
}
